src/base/bittorrent/session.cpp

```diff
@@ -4355,6 +4356,9 @@ void Session::handleAlert(const lt::alert *a)
         case lt::socks5_alert::alert_type:
             handleSocks5Alert(static_cast<const lt::socks5_alert *>(a));
             break;
+        case lt::piece_finished_alert::alert_type:
```

Comment:
CC: @arvidn

Reply:
If it is too performance intensive, it could be gated behind a setting that's disabled by default, and the setting label could warn the user. For instance, `Enable pieces map (WARNING: resource intensive): [✓]`

Reply:
IIRC, it's advanced variant of existing "pieces bar", isn't it?
Anyway, why not update it in the similar way as many other GUI elements, i.e. discrete?

Reply:
regarding performance:

1. since this is only enabling `piece_progress_notification` you won't be receiving the `block_*_alert`s. Those are the real fire hose. However, in earlier versions of libtorrent (1.1), `piece_progress_notification` and `block_progress_notification` were the same alert category (`progress_notification`), so if you needed one you got the other. However, this shouldn't be a problem anymore. Now you'll just get one alert per *piece*, which may be a lot, but not close to an alert for every block.
2. A more efficient way to handle these updates would be to coalesce all `piece_finished_alert`s received in the same `pop_alerts()` call, and then just update the GUI *once* with all the new pieces it just learned about.

Reply:
>A more efficient way to handle these updates would be to coalesce all `piece_finished_alerts` received in the same `pop_alerts()` call, and then just update the GUI once with all the new pieces it just learned about.

It can still be expensive a lot if download speed is big. Why waste resources on this? What is the practical use of this? Isn't it enough to update it once every one or two seconds?

Reply:
>@Chocobo1, what do you think?

I agree with you, it shouldn't be real-time updated by default.
It can be toggled real-time by an advanced settings but I would suggest we keep it simple in this first PR and reconsider it later.

Reply:
the flip side is if you have a lot of torrents that aren't moving at all, updating them anyway, on a schedule, is wasteful.

Reply:
>the flip side is if you have a lot of torrents that aren't moving at all, updating them anyway, on a schedule, is wasteful.

Such things as piece map is supposed to be shown only for single torrent at once.